### ISSUE TITLE
Reduce load extension non-determinism via metadata predicate specificity

### ISSUE THREAD

**Issue by unscriptable** [RaveJS/rave#14]
### Status quo

Currently, rave allows load extensions via an adhoc mechanism.  If an extension wants to extend the common loader, it may do so i n any way it likes.  The ES6-prescribed way to override a loader is to use method replacement (aka "AOP around advice") with a predicate that determines whether to use a custom loader hook or defer to the existing loader hook.  
### The problem

The problem with this approach becomes evident when there are multiple custom loader hooks being applied.  Since loader hooks are typically implemented within modules -- and since modules load asynchronously -- there is no way to ensure that the multiple loader hooks will load in a deterministic order.  In short: it's not possible for the end user to predict or control the priority of loader hooks without manual configuration.

**_We would like to eliminate all configuration, if possible, so we need an alternative.**_
### Proposed solution

One way to provide determinism is to rank loader hooks. One way to assign rank is via specificity (similar to CSS!).  The more specific loader hooks should have higher priority.  For instance, a loader hook that loads text files is less specific than a loader hook that is specialized for markdown-type text files.

ES6 does not specify any way to provide specificity.  Method replacement and predicate functions don't expose any usable information (and AST is too expensive and problematic).  However, given our experience with AMD and CommonJS loaders, we can hypothesize a reasonable set of conditions that developers will be testing in their predicates.  If we expose a structured way for extension authors to take advantage of these conditions as metadata, we have a way to measure specificity.

Here's the list that rave will start with.  We can add more or adjust the ranking of these over time.

Predicates ordered by rank:
1. **`package`**: string. Extension authors should specify a single package to which their loader hooks will apply.  Lack of a package will indicate that the hooks will apply to all packages.  In the future, we could support wildcards and comma-separated package names.  The author may specify a custom function to determine if a module name belongs to package. _Specificity_: `package` has highest rank, fewer packages indicates higher specificity.
2. **`pattern`**: string|RegExp. Extension authors may also specify a pattern that must match on the normalized module name.  If the pattern is a string, it is converted to a RegExp via `new Regexp(pattern)`.  In the future, we could allow multiple patterns or parse a pattern's specificity by the number of `|`.  AMD plugin extensions will use `pattern`.  _Specificity_: `pattern` has lower rank than `package`, the existence of a pattern is more specific than not.  
3. **`extensions`**: Array<string>.  Extension authors may wish to target file extensions.  This is similar to how node.js works and is in line with how the cujoJS team has been thinking over the past few years.  _Specificity_: `extensions` has lower rank than `pattern`, fewer extensions indicates more specificity.
4. **`predicate`**: function (load: {{ name: string }}): boolean. When the other metadata predicates are insufficient, the extension author may write a custom predicate function.  The predicate function takes an object similar to the `load` object passed to the loader hooks.  In all cases, the `load` object will have a `name` property whose value is the normalized name of the module being loaded.  To help provide determinism, we rank the custom predicate _last_.  _Specificity_: `predicate` has the lowest rank, the existence of a predicate is more specific than the lack of one.
### Examples

Examples of metadata exposed by a load extension:

``` js
// A load extension for a specific package, "frozen" that requires multiple AMD plugins.
// This load extension will trump all other general-purpose AMD plugins.
{
    package: 'frozen',
    pattern: '\\b!\\b'
}
```

``` js
// A load extension for a general-purpose AMD plugin loader.
// This load extension will not apply to modules in the "frozen" package.
{
    pattern: /\b!\b/
}
```

``` js
// A load extension for a general-purpose text loader.
{
    extensions: ['.text', '.txt', '.html', '.htm', '.css', '.md', '.mdown', '.markdown']
}
```

``` js
// A load extension for a general-purpose stylesheet loader.
// This extension is more specific than the text loader above.
{
    extensions: ['.css']
}
```

**Comment by unscriptable** [comment]
Got this working internally to rave.  Still need to do the following:
- [ ] modify how rave extensions extend the loader
- [ ] create rave-load-json since the internal json loader doesn't use metadata to find modules
- [ ] modify the existing rave extensions.

Q: how do we coordinate the release of rave and rave extensions when breaking API changes are made?

**Comment by unscriptable** [comment]
This has landed in 0.1.0.
